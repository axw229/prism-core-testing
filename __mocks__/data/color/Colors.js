// @flow
import cloneDeep from 'lodash/cloneDeep'
import sample from 'lodash/sample'

import type { ColorSetPayload, Color } from '../../../src/shared/types/Colors'

export const Colors: ColorSetPayload = {
  "Red": [
    [
      {
        "colorNumber": "6561",
        "coordinatingColors": {
          "coord2ColorId": 1926,
          "coord1ColorId": 2250,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2248",
        "isExterior": true,
        "isInterior": true,
        "name": "Teaberry",
        "rgb": 15454683,
        "lrv": 68.551,
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 235,
        "green": 209,
        "blue": 219,
        "hue": 0.935897435897436,
        "saturation": 0.39393939393939387,
        "lightness": 0.8705882352941177,
        "hex": "#ebd1db",
        "isDark": false,
        "cssrgb": "rgb(235,209,219)",
        "storeStripLocator": "102-C1",
        "similarColors": [
          "2255",
          "2262",
          "2269",
          "2662",
          "2535",
          "2276",
          "2658",
          "1974",
          "2543",
          "2531"
        ]
      },
      {
        "colorNumber": "6568",
        "coordinatingColors": {
          "coord2ColorId": 11244,
          "coord1ColorId": 2687,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2255",
        "isExterior": true,
        "isInterior": true,
        "name": "Lighthearted Pink",
        "rgb": 15586781,
        "lrv": 70.758,
        "brandedCollectionIds": [
          31797,
          31791
        ],
        "brandedCollectionNames": [
          "Pottery Barn Fall/Winter 2018",
          "Pottery Barn Fall/Winter 2018"
        ],
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 237,
        "green": 213,
        "blue": 221,
        "hue": 0.9444444444444445,
        "saturation": 0.3999999999999999,
        "lightness": 0.8823529411764706,
        "hex": "#edd5dd",
        "isDark": false,
        "cssrgb": "rgb(237,213,221)",
        "storeStripLocator": "103-C1",
        "similarColors": [
          "2248",
          "2262",
          "2269",
          "2276",
          "1974",
          "1981",
          "2658",
          "1995",
          "2662",
          "1988"
        ]
      }
    ]
  ]
}

export function getColors (): ColorSetPayload {
  return cloneDeep(Colors)
}

export function getColor (): Color | void {
  const colors = getColors()
  const fam1 = Object.keys(colors)[0]
  if (colors[fam1]) {
    return sample(colors[fam1][0])
  }
}
