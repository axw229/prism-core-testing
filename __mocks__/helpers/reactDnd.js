import React from 'react'
import { shallow, mount } from 'enzyme'
import { DndProvider } from 'react-dnd-cjs'
import HTML5Backend from 'react-dnd-html5-backend-cjs'

// Create IntlProvider to retrieve React Intl context
const wrappingComponentProps = {
  backend: HTML5Backend
}

export function mountWithDnd (node: React.ReactElement, options = {}) {
  const _options = options || {}

  return mount(node, {
    wrappingComponent: DndProvider,
    wrappingComponentProps,
    ..._options
  })
}

export function shallowWithDnd (node: React.ReactElement, options = {}) {
  const _options = options || {}

  return shallow(node, {
    wrappingComponent: DndProvider,
    wrappingComponentProps,
    ..._options
  })
}

export default shallowWithDnd
