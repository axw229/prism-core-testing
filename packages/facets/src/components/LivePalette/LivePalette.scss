.prism-live-palette {
  display: flex;
  background-color: var(getThemeColor('white'));
  padding: $livePalette__chip-side-margin;
  font-family: var(getVarName('typography.titleFontFamily'));
  @include respond-above(sm) {
    border: 1px solid $color-grey;
  }

  &-simple {
    border: none;
  }

  &__header {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    display: none;
    @include respond-above(sm) {
      display: flex;
    }
    &__name {
      display: none;
      text-align: center;
      @include respond-above(sm) {
        display: block;
        margin: auto 0.5em;
      }
    }
    &__compare-button{
      &--hide{
        display: none;
      }
      padding: 0.5em 1em;
      margin-right: 5px;
      font-family: var(getVarName('typography.titleFontFamily'));
      font-size: .8em;
      outline:none;
      text-transform: var(getVarName('typography.buttonTextTransform'));
      font-weight: var(getVarName('typography.buttonFontWeight'));
      border: 1px solid var(getThemeColor('buttonBorder'));
      background-color: var(getThemeColor('buttonBgColor'));
      color: var(getThemeColor('buttonColor'));

      &:hover{
        border: 1px solid var(getThemeColor('buttonHoverBorder'));
        background-color: var(getThemeColor('buttonHoverBgColor'));
        color: var(getThemeColor('buttonHoverColor'));
        cursor: pointer;
      }
    }
  }

  &__height {
    @include respond-above(sm) {
      height: 4rem !important;
    }
    @include respond-above(md) {
      height: 4.625rem !important;
    }
  }

  &__slot {
    $lp-icon-size: 24px;

    width: inherit;
    height: 100%;

    &:focus {
      outline-style: solid;
      outline-color: var(getThemeColor('secondary'));
      outline-width: 2px;
    }

    &--empty {
      display: flex;
      justify-content: center;
      align-items: center;
      width: inherit;
      height: 100%;
      cursor: default;
      background-color: var(getThemeColor('light'));
    }

    &__label {
      display: none;
      font-size: #{(9/16)}rem;
      text-align: center;

      @include respond-above(md) {
        display: block;
      }

      &--text {
        width: 9rem;
        margin: 0.2rem;
        font-family: var(getVarName('typography.bodyFontFamily'));
        font-style: var(getVarName('typography.titleFontStyle'));
        font-size: .875rem;
        line-height: 1.2rem;
      }
    }

    &__icon {
      width: $lp-icon-size !important;
      height: $lp-icon-size !important;
      margin: 5px;
      color: var(getThemeColor('light.contrast'));

      &--empty { opacity: .3 }
    }

    &__button {
      position: absolute;
      top:50%;
      left:50%;
      transform: translate(-50%, -50%);
      outline:none;
    }
  }

  &__info {
    margin: 0 .125rem !important;
    font-size: 1.4rem !important;

    &--outlined {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 1rem;
      height: 1rem;
      outline: 2px solid;

      &-icon { font-size: .8rem !important }
    }

    &:focus {
      outline-style: solid;
      outline-color: var(getThemeColor('secondary'));
      outline-width: 2px;
    }
  }

  &__trash {
    @include respond-above(sm) {
      margin-left: .25rem !important;
    }

    &:focus {
      outline-style: solid;
      outline-color: var(getThemeColor('secondary'));
      outline-width: 2px;
    }

    &--icon { font-size: 20px }
  }

  &__name-number {
    display: flex;
    flex-direction: column-reverse;
  }

  &__color-number {
    display: block;
    font-size: .8rem;
    font-weight: normal;
    color: inherit;
    white-space: nowrap;
  }

  &__color-name {
    display: block;
    font-size: .8rem;
    color: inherit;
    @include respond-above(sm) {
      font-weight: 600;
    }
  }

  &__color-description {
    clip: rect(1px, 1px, 1px, 1px); // For screen readers only
    font-size: 0.7em;
    display: none;
    position: absolute;
    bottom: 0.5em;
    color: inherit;
    @include respond-above(sm) {
      display: block;
    }
  }

  &-house-shaped {
    &__name-number {
      display: flex;
      flex-direction: column-reverse;
      line-height: 1.2rem;
    }

    &__color-name {
      font-family: var(getVarName('typography.bodyFontFamily'));
      font-style: italic;
      white-space: nowrap;
    }

    &__color-number {
      font-size: .875rem;
      line-height: .9rem;
      white-space: nowrap;
    }

    &__height {
      @include respond-below(sm) {
        height: 5.6rem !important;
      }
    }
  }
}
