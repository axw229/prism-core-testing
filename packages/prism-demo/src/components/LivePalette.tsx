import React, { useContext } from 'react'
import Prism, { LivePalette, ColorsIcon } from '@prism/toolkit'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faInfo, faTrash } from '@fortawesome/pro-solid-svg-icons'
import { faPlusCircle } from '@fortawesome/pro-light-svg-icons'
import { filter, values } from 'lodash'
import { withColorData } from './withColorData'
import { Color } from '../types'
import Context from "../context"

const LivePaletteWrapper = ({ colors }: { colors: Color[]}) => {
  const { lpColors, resetLpColors } = useContext(Context)

  return (
    <Prism>
      <LivePalette
        addButtonRenderer={colors => (
          <button className='h-full' onClick={() => { console.log('"Add a Color" button triggered') }} style={{ width: 'inherit' }}>
            <div className={`flex justify-center items-center ${colors.length === 0 ? '' : 'flex-col'}`}>
              <FontAwesomeIcon icon={faPlusCircle} size='lg' />
              <p className={`hidden md:block ${colors.length === 0 ? 'm-1.5 text-left' : 'm-1'} text-2xs`}>
                {colors.length === 0 ? <>FIND COLORS IN<br /> THE DIGITAL COLOR WALL</> : 'ADD A COLOR'}
              </p>
            </div>
          </button>
        )}
        // @ts-ignore
        colors={lpColors}
        deleteButtonRenderer={({ name }, onClick) => (
          <button className='md:ml-1 ring-primary focus:outline-none focus-visible:ring-2' onClick={onClick}>
            <FontAwesomeIcon icon={faTrash} style={{ fontSize: '20px'}} />
          </button>
        )}
        detailsButtonRenderer={({ coordinatingColors, name }) => (
          <button
            className='mx-0.5 ring-primary focus:outline-none focus-visible:ring-2'
            onClick={() => { console.log('Details button triggered') }}
          >
            <ColorsIcon
              className='w-5 h-5'
              hexes={filter(colors, (c) => values(coordinatingColors).some((id) => id === c.id)).map((c) => c.hex)}
              infoIcon={<FontAwesomeIcon icon={faInfo} style={{ margin: '0 0.25rem' }} />}
            />
          </button>
        )}
        emptySlotRenderer={() => (
          <div className='flex h-full items-center justify-center opacity-60 bg-light cursor-default' style={{ width: 'inherit' }}>
            <FontAwesomeIcon icon={faPlusCircle} size='lg' />
          </div>
        )}
        labelRenderer={({ brandKey, colorNumber, name }) => (
          <div className='text-xs'>
            <p className='whitespace-nowrap'>{`${brandKey} ${colorNumber}`}</p>
            <p className='whitespace-nowrap md:font-bold'>{name}</p>
          </div>
        )}
        onColorsChanged={(colors) => resetLpColors(colors)}
      />
    </Prism>
  )
}

export default withColorData(LivePaletteWrapper)
