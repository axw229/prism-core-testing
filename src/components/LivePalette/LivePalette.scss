.prism-live-palette {
  max-width: 990px;
  overflow: hidden;
  background-color: getVar( "colors.white" );
  border: 1px solid getVar( "colors.grey" );
  padding: $livePalette__chip-side-margin;

  &__list {
    display: flex;
    flex-grow: 1;
    margin: 0;
  }

  &__slot {
    @include chip-attributes($distribution: even);
    background-color: #ebebeb;
    position: relative;
    cursor: pointer;
    transition: width .6s ease-out, opacity .2s ease-out;
    height: 0;
    overflow: hidden;

    &--empty,
    &--add {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: center;
    }

    &--empty {
      cursor: default;
    }

    &--add {
      background-color: getVar( "colors.white" );
    }

    &--add-empty {
      width: 30%;

      .prism-live-palette__slot__copy {
        width: 75%;
        top: 50%;
        left: 5em;
        bottom: auto;
        text-align: left;
        transform: translateY(-50%);
      }

      .prism-live-palette__icon {
        left: 0.5em;
      }
    }

    &--active {
      width: 25%; // TODO: calculate actual width
      opacity: 1;

      .prism-live-palette__color-details {
        display: block;
      }
    }

    &--removing {
      opacity: 0.01;
      transition: 700ms;
    }

    &__copy {
      position: absolute;
      display: block;
      bottom: 0.75em;
      font-size: 0.7rem;
      text-decoration: none;
    }
  }

  &__icon {
    top: 50%;
    position: absolute;
    color: #000000;
    transform: translateY(-50%);
  }

  &__trash {
    position: absolute;
    top: 0.5em;
    right: 0.5em;
    color: inherit;
    background: inherit;
    border: none;
    box-shadow: none;
  }

  &__color-details {
    height: auto;
    position: static;
    padding: 0.5em;
    display: none;

    &--dark {
      color: getVar( "colors.white" );
    }
  }

  &__color-number {
    font-size: 0.8rem;
    display: block;
    color: inherit;
  }

  &__color-name {
    font-size: 0.8rem;
    font-weight: bold;
    display: block;
    color: inherit;
  }

  &__color-description {
    clip: rect(1px, 1px, 1px, 1px); // For screen readers only
    font-size: 0.7em;
    display: block;
    position: absolute;
    bottom: 0.5em;
    color: inherit;
  }
}
