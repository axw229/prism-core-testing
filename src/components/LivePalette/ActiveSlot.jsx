// @flow
import React, { useState, useRef } from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useDrag, useDrop } from 'react-dnd'
import { fullColorNumber } from '../../shared/helpers/ColorUtils'
import { type Color } from '../../shared/types/Colors'

import { remove, activatePreview } from '../../store/actions/live-palette'

import { DRAG_TYPES } from 'constants/globals'

type Props = {
  active: boolean,
  color: Color,
  index: Number,
  remove: Function,
  onClick: Function,
  moveColor: Function
}

export function ActiveSlot (props: Props) {
  const ref = useRef(null)
  const { color, active, index } = props
  const [isDeleting, setDeletingStatus] = useState(false)

  const ACTIVE_CLASS = 'prism-live-palette__slot--active'
  const REMOVAL_CLASS = 'prism-live-palette__slot--removing'

  const LIGHT_DARK_CLASS = color.isDark ? 'prism-live-palette__color-details--dark' : 'prism-live-palette__color-details--dark-color'

  const [, drop] = useDrop({
    accept: DRAG_TYPES.SWATCH,
    hover (item, monitor) {
      if (!ref.current) {
        return
      }

      const dragIndex = item.index
      const hoverIndex = index

      // prevent items from replacing themself
      if (dragIndex === hoverIndex) {
        return
      }

      // determine rectangle on screen
      const hoverBoundingRect = ref.current.getBoundingClientRect()

      // get horizontal middle
      const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2

      // determine mouse position
      const clientOffset = monitor.getClientOffset()

      // get pixels to the left
      const hoverClientX = clientOffset.x - hoverBoundingRect.left

      // only perform the move when the mouse has crossed half of the items height
      if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
        return
      }
      if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
        return
      }

      // perform the move action
      props.moveColor(dragIndex, hoverIndex)

      item.index = hoverIndex
    }
  })

  const activateSlot = () => {
    // only trigger the onClick and activating the swatch if it's not already active.
    // this should prevent an additional re-render since clicking on the move icon also triggers the click
    // on the swatch itself
    if (!props.active && props.onClick) {
      props.onClick(color)
    }
  }

  const remove = () => {
    setDeletingStatus(true)
    setTimeout(() => {
      props.remove(color.id)
    }, 200)
  }

  const [{ isDragging }, drag] = useDrag({
    item: { type: DRAG_TYPES.SWATCH, color, index },
    collect: monitor => ({
      isDragging: monitor.isDragging()
    })
  })
  const opacity = isDragging ? 0.33 : 1

  drag(drop(ref))

  return (
    <div ref={ref} className={`prism-live-palette__slot ${(active ? ACTIVE_CLASS : '')} ${(isDeleting ? REMOVAL_CLASS : '')}`} style={{ backgroundColor: color.hex, opacity }} onClick={activateSlot} onKeyDown={activateSlot} role='button' tabIndex='-1'>
      <div className={`prism-live-palette__color-details ${LIGHT_DARK_CLASS}`}>
        <span className='prism-live-palette__color-number'>{fullColorNumber(color.brandKey, color.colorNumber)}</span>
        <span className='prism-live-palette__color-name'>{ color.name }</span>
        <span className='prism-live-palette__color-description'>{ color.description.join(', ') }</span>
        <button className='prism-live-palette__trash' onClick={remove}><FontAwesomeIcon icon={['fa', 'trash']} size='1x' /></button>
      </div>
    </div>
  )
}

const mapDispatchToProps = (dispatch) => {
  return {
    remove: (colorId) => {
      dispatch(remove(colorId))
    },
    activatePreviewColor: (color) => {
      dispatch(activatePreview(color))
    }
  }
}

export default connect(null, mapDispatchToProps)(ActiveSlot)
