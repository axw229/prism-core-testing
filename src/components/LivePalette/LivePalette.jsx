// @flow
import type { Color } from '../../shared/types/Colors'

import React, { PureComponent } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { FormattedMessage } from 'react-intl'
import { connect } from 'react-redux'
import without from 'lodash/without'
import times from 'lodash/times'
import flatMap from 'lodash/flatMap'
import intersection from 'lodash/intersection'
import update from 'immutability-helper'
import { Link } from 'react-router-dom'
import LivePaletteModal from './LivePaletteModal'
import store from '../../store/store'
import { LP_MAX_COLORS_ALLOWED, MIN_COMPARE_COLORS_ALLOWED } from 'constants/configurations'

import { activate, reorder, toggleCompareColor, cancel, empty } from '../../store/actions/live-palette'
import { arrayToSpacedString } from '../../shared/helpers/StringUtils'

import { varValues } from 'variables'

import EmptySlot from './EmptySlot'
import ActiveSlot from './ActiveSlot'

import './LivePalette.scss'

type Props = {
  colors: Array<Color>,
  activateColor: Function,
  reorderColors: Function,
  toggleCompareColor: Function,
  activeColor: Color,
  removedColor: Color,
  cancel: Function,
  empty: Function
}

type State = {
  spokenWord: string,
  isCompareColor: boolean
}

export class LivePalette extends PureComponent<Props, State> {
  state = {
    spokenWord: '',
    isCompareColor: false
  }

  pendingUpdateFn: any
  requestedFrame: number | void
  activeSlotRef: ?RefObject = void (0)

  constructor (props: Props) {
    super(props)

    this.activeSlotRef = React.createRef()
  }

  componentDidMount () {
    store.subscribe(() => {
      const { lp } = store.getState()
      window.localStorage.setItem('lp', JSON.stringify(lp))
    })
  }
  // $FlowIgnore
  componentDidUpdate (prevProps, prevState) {
    let spokenWord: Array<string> = []
    const prevColor = prevProps.activeColor
    const curColor = this.props.activeColor

    if (this.props.removedColor && prevProps.removedColor !== this.props.removedColor) {
      spokenWord.push(`${this.props.removedColor.name} has been removed.`)
    }

    if (curColor && prevColor !== curColor) {
      spokenWord.push(`${curColor.name} has been activated.`)

      // POC AHEAD
      // TODO: This depends on manual intervention to remove unhelpful color families. Type of color (red, yellow, etc) should be
      // determined programmatically based on color values.
      if (prevColor) {
        let mainFam = without(intersection(prevColor.colorFamilyNames, curColor.colorFamilyNames), 'Timeless Color', 'Historic Color', 'White & Pastel')[0]

        if (mainFam) {
          let prevLightness = prevColor.lightness
          let curLightness = curColor.lightness

          let brightnessDifference = (curLightness > prevLightness ? 'lighter' : curLightness < prevLightness ? 'darker' : void (0))

          if (brightnessDifference) {
            spokenWord.push(`${curColor.name} is a ${brightnessDifference} ${mainFam} than ${prevColor.name}`)
          }
        }
      }
      // END POC
    }

    if (spokenWord.length) {
      this.setState({ // eslint-disable-line react/no-did-update-set-state
        spokenWord: arrayToSpacedString(spokenWord)
      })
    }
  }

  toggleCompareColor = () => {
    this.setState({ isCompareColor: !this.state.isCompareColor })
    this.props.toggleCompareColor()
  }

  render () {
    const { colors, activeColor, cancel, empty } = this.props

    const { spokenWord, isCompareColor } = this.state
    // TODO: abstract below into a class method
    // calculate all the active slots
    const activeSlots = colors.map((color, index) => {
      if (color && index < LP_MAX_COLORS_ALLOWED) {
        return (<ActiveSlot
          ref={this.activeSlotRef}
          node={this.activeSlotRef} // passing the ref down as a prop so DnD has access to the DOM element
          index={index}
          key={color.id}
          color={color}
          onClick={this.activateColor}
          moveColor={this.moveColor}
          active={(activeColor.id === color.id)}
          isCompareColor={isCompareColor}
        />)
      }
    })

    // after determining active slots, determine how many empty ones there should be
    let disabledSlots = []
    const additionalSlots = (LP_MAX_COLORS_ALLOWED - 1) - activeSlots.length
    if (additionalSlots > 0) {
      disabledSlots = times(additionalSlots, (index) => <EmptySlot key={index} />)
    }

    const ADD_COLOR_TEXT = (colors.length) ? 'ADD_A_COLOR' : 'FIND_COLORS_IN_CW'
    const COLOR_TRAY_CLASS_MODIFIERS = (colors.length) ? 'add' : 'add-empty'
    return (
      <div className='prism-live-palette'>
        <LivePaletteModal cancel={cancel} empty={empty} isActive={colors.length > LP_MAX_COLORS_ALLOWED} />
        <div className='prism-live-palette__header'>
          <span className='prism-live-palette__header__name'><FormattedMessage id='PALETTE_TITLE' /></span>
          {colors.length >= MIN_COMPARE_COLORS_ALLOWED && <button className='prism-live-palette__header__compare-button' onClick={this.toggleCompareColor}>Compare Color</button>}
        </div>
        <div className='prism-live-palette__list'>
          {activeSlots}
          {colors.length < LP_MAX_COLORS_ALLOWED && <Link to={`/active/color-wall`} className={`prism-live-palette__slot prism-live-palette__slot--${COLOR_TRAY_CLASS_MODIFIERS}`}>
            <FontAwesomeIcon className='prism-live-palette__icon' icon={['fal', 'plus-circle']} size='2x' color={varValues.colors.swBlue} />
            <FormattedMessage id={ADD_COLOR_TEXT}>
              {(msg: string) => <span className='prism-live-palette__slot__copy'>{msg}</span>}
            </FormattedMessage>
          </Link>}
          {disabledSlots}
        </div>
        {/* This will speak the current and removed color, as well as some color-delta info. */}
        <aside aria-live='assertive' className='prism-live-palette__color-description'>{spokenWord}</aside>
      </div>
    )
  }

  activateColor = (color: Color) => {
    this.props.activateColor(color)
  }

  scheduleUpdate = (updateFn: Function) => {
    this.pendingUpdateFn = updateFn

    if (!this.requestedFrame) {
      this.requestedFrame = window.requestAnimationFrame(this.drawFrame)
    }
  }

  drawFrame = () => {
    const sortedColorsById = update([], this.pendingUpdateFn)

    // trigger the reordering via redux
    this.props.reorderColors(sortedColorsById)

    this.pendingUpdateFn = undefined
    this.requestedFrame = undefined
  }

  moveColor = (originColorId: Number, destinationColorId: Number) => {
    const { colors } = this.props
    const colorsByIndex = flatMap(colors, color => color.id) // creates an array of only all color ids
    const originIndex = colorsByIndex.indexOf(originColorId) // get the index of the origin color
    const destIndex = colorsByIndex.indexOf(destinationColorId) // get the index of the dest color

    // shuffle the origin with the dest
    const from = colorsByIndex.splice(originIndex, 1)[0]
    colorsByIndex.splice(destIndex, 0, from)

    // schedule the rearrangement of a swatch with the browser
    this.scheduleUpdate({
      $push: colorsByIndex
    })
  }
}

const mapStateToProps = (state, props) => {
  const { lp } = state
  return {
    colors: lp.colors,
    activeColor: lp.activeColor,
    // previousActiveColor: lp.previousActiveColor,
    removedColor: lp.removedColor
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    toggleCompareColor: () => {
      dispatch(toggleCompareColor())
    },
    activateColor: (color) => {
      dispatch(activate(color))
    },
    reorderColors: (colors) => {
      dispatch(reorder(colors))
    },
    cancel: () => {
      dispatch(cancel())
    },
    empty: () => {
      dispatch(empty())
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(LivePalette)
