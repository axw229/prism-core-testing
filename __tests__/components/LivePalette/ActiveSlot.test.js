import React from 'react'
import { ActiveSlot } from 'src/components/LivePalette/ActiveSlot'
import * as Colors from '__mocks__/data/color/Colors'
import { mountWithDnd } from '__mocks__/helpers/reactDnd'

const colors = Colors.getColor()
const defaultProp = {
  color: colors,
  connectDragSource: jest.fn((el) => el),
  connectDropTarget: jest.fn((el) => el)
}
/** @see https://github.com/react-dnd/react-dnd/issues/1494 */
const createActiveSlot = (props) => {
  return mountWithDnd(<ActiveSlot {...defaultProp} {...props} />)
}

// rendering testing
describe('rendering testing for activeslot compoenent', () => {
  it('active slot component should rendering correctly with colors', () => {
    const wrapper = createActiveSlot()
    expect(wrapper.find('.prism-live-palette__slot').exists()).toBe(true)
    expect(wrapper.find('.prism-live-palette__color-details').exists()).toBe(true)
    expect(wrapper.find('.prism-live-palette__color-number').exists()).toBe(true)
    expect(wrapper.find('.prism-live-palette__color-name').exists()).toBe(true)
    expect(wrapper.find('.prism-live-palette__color-description').exists()).toBe(true)
    expect(wrapper.find('.prism-live-palette__trash').exists()).toBe(true)
  })
})

// event test
describe('Event testing for activeslot component', () => {
  it('onClick and activating the swatch should be invoked if its not already active', () => {
    const spy = jest.fn()
    const wrapper = createActiveSlot({ active: false, onClick: spy })
    wrapper.find('.prism-live-palette__slot').simulate('click')
    expect(spy).toHaveBeenCalledWith(colors)
  })

  it('onClick and activating the swatch should not be invoked if its already active', () => {
    const spy = jest.fn()
    const wrapper = createActiveSlot({ active: true, onClick: spy })
    wrapper.find('.prism-live-palette__slot').simulate('click')
    expect(spy).not.toHaveBeenCalledWith(colors)
  })

  it('remove function should be invoked if click remove button', () => {
    const spy = jest.fn()
    const wrapper = createActiveSlot({ remove: spy })
    wrapper.find('.prism-live-palette__trash').simulate('click')
    setTimeout(() => { expect(spy).toHaveBeenCalledWith(colors.id) }, 300)
  })
})
