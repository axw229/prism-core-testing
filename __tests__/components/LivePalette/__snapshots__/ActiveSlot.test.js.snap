// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing for activeslot component  ActiveSlot component should match snapshot 1`] = `
<Memo()
  backend={[Function]}
>
  <RootFinder>
    <ActiveSlot
      color={
        Object {
          "blue": 219,
          "brandKey": "SW",
          "colorFamilyNames": Array [
            "Red",
          ],
          "colorNumber": "6561",
          "coordinatingColors": Object {
            "coord1ColorId": 2250,
            "coord2ColorId": 1926,
            "whiteColorId": 2681,
          },
          "cssrgb": "rgb(235,209,219)",
          "description": Array [
            "Subdued",
            "Muted",
            "Brilliant",
            "Radiant",
          ],
          "green": 209,
          "hex": "#ebd1db",
          "hue": 0.935897435897436,
          "id": "2248",
          "isDark": false,
          "isExterior": true,
          "isInterior": true,
          "lightness": 0.8705882352941177,
          "lrv": 68.551,
          "name": "Teaberry",
          "red": 235,
          "rgb": 15454683,
          "saturation": 0.39393939393939387,
          "similarColors": Array [
            "2255",
            "2262",
            "2269",
            "2662",
            "2535",
            "2276",
            "2658",
            "1974",
            "2543",
            "2531",
          ],
          "storeStripLocator": "102-C1",
        }
      }
      connectDragSource={[MockFunction]}
      connectDropTarget={[MockFunction]}
    >
      <div
        className="prism-live-palette__slot  "
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        style={
          Object {
            "backgroundColor": "#ebd1db",
            "opacity": 1,
          }
        }
        tabIndex="-1"
      >
        <div
          className="prism-live-palette__color-details prism-live-palette__color-details--dark-color"
        >
          <span
            className="prism-live-palette__color-number"
          >
            SW 6561 
          </span>
          <span
            className="prism-live-palette__color-name"
          >
            Teaberry
          </span>
          <span
            className="prism-live-palette__color-description"
          >
            Subdued, Muted, Brilliant, Radiant
          </span>
          <button
            className="prism-live-palette__trash"
            onClick={[Function]}
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "fa",
                  "trash",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size="1x"
              spin={false}
              symbol={false}
              title=""
              transform={null}
            />
          </button>
        </div>
      </div>
    </ActiveSlot>
  </RootFinder>
</Memo()>
`;
